modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java

```diff
@@ -264,25 +266,7 @@ protected void onNodeLeft(UUID evtNodeId) {
     protected void enqueue(Collection<?> col) {
         assert Thread.holdsLock(this);
 
-        if (limitDisabled) {
-            queue.add((Collection<R>)col);
-
-            cnt.addAndGet(col.size());
-        }
-        else {
-            if (capacity >= col.size()) {
-                queue.add((Collection<R>)col);
-                capacity -= col.size();
-
-                cnt.addAndGet(col.size());
-            }
-            else if (capacity > 0) {
-                queue.add(new ArrayList<>((Collection<R>)col).subList(0, capacity));
-                capacity = 0;
-
-                cnt.addAndGet(capacity);
-            }
-        }
+        queue.add((Collection<R>)col);
```

Comment:
Why are you remove limit tracking from here? 
Seems, onDone() + cancel() methods could be called when the limit was reached.
onDone() will prevent requesting next pages (loadPage()), and cancel() will free resources on map nodes.

Reply:
Hi! Sorry, could you please clarify which case do you mean? How does the cancel() invoke? 

If I understand you correctly, you told that empty collection (that will be inserted to the queue after limit exceed) will trigger cancel. But actually it's not true. In this case we just check that it's empty and ask for new page [1]. So nobody will send cancel().

You can run my test "GridCacheTextQueryPagesTest" from this PR on master code base - it will fail. It checks number of pages to be loaded during a query.

[1] https://github.com/apache/ignite/blob/7faa57c49118285b830fcf614a12abda65116f51/modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java#L213

Reply:
Thanks for clarification, now understood your note. You're correct that it will work safe on current state of master. It's because we don't have any reduce of filter operations on received pages. But I did it by 2 reasons: 

1. In parallel activity I add MergeSort algorithm, it merging data from different pages. So there are cases when we get a data from single node, and it covers a specified limit. But actually we need another page from other nodes to compare. + Maybe there could be additional cases in future for other reduce or filtering operations. 

2. I don't like idea to make `enqueue()` responsible for finishing future, as there is already a working mechanism of `onDone()` and `cancel()` based on `hasNext()` result. Also now `enqueue()` creates a new collection from sublist in case of limit applying on input collection, there is not much overhead but it can be avoid by the same existing mechanism. I believe the less code the better. So I decide remove this duplicated mechanism.